### DISCUSSION TITLE
From a scripting point of view, given a block, whats the easiest way to find the associated blocks?

### DISCUSSION THREAD

**Discussion by mikekidner** [gaphor/gaphor#2766] [Q&A] [answered]
I have created a simple FMEA profile by stereotyping the basic SysML block to FailureMode and FailureCause, these can be linked by associations to each other and the block of interest.
Now, I'm trying to write a script to extract the linked items to form a row of a typical FMECA table.

Given a block, what is the easiest way to follow its connections to other blocks? My scripting so far seems to show up many empty collections and nested-nested Ends and Members etc 

I get all the blocks stereotyped as FailureMode like this: 
```
filterFailureModes = lambda e: isinstance(e, UML.Class) and   'Failure Mode' in list(e.appliedStereotype[:].classifier.name[:])
        self.failureModeObjs = [a for a in self.factory.select(filterFailureModes)]
```
I can get my FailureCause blocks in the same way.

But now I want to follow the associations between them.

**Comment by mikekidner** [comment]
I've tried starting from getting the associations first and then trying to get the blocks, but while I could find references to the named ends (`.member` and `.memberEnd`) of the association, and the type of element at each end (`.endType`) , there was no reference to the actual connected element.
I was expecting the associations would have `.supplierDependency` and `.clientDependency` properties which would point to the connected blocks - but these properties are empty.


In the UML model there is a comment that says "fix associations"; 
<img width="612" alt="Screen Shot 2023-10-12 at 11 47 34 am" src="https://github.com/gaphor/gaphor/assets/50702106/e71d698d-2ae8-4c8a-8a18-3023d7ded555">

Does anyone know what needs to be fixed? 
Does it relate to being able to trace an association to the elements it connects to?

**Comment by amolenaar** [comment, marked as answer]
Indeed, `endType` (or just `type`) points to the type of an association end. Association ends (`gaphor.UML.Property` instances) always have a `memberEnd` property that points to the association. If an association is *navigable*, you should see the property as part of the `Type`'s `ownedAttribute` property (given the type has an `ownedAttribute` property).

For background, here's the [`set_navigability`](https://github.com/gaphor/gaphor/blob/main/gaphor/UML/recipes.py#L282-L343) function.

**Reply by mikekidner** [reply]
Thanks @amolenaar adding navigability helps, so now it looks like :

<img width="563" alt="Screen Shot 2023-10-16 at 11 40 07 am" src="https://github.com/gaphor/gaphor/assets/50702106/36618000-5baa-4587-8ee1-cc0b96884381">

which allows either a system, a failure mode or a failure cause to be traced to the related items.

```
   fmea.getFailureCause('sys1 Fail Mode')

    for x in fmea.failureCauseObjs:
        print(x.qualifiedName)
        for oa in x.ownedAttribute:
            print(oa.type.name)
``` 
Which yields:
```
TEST/test_fmeaParser.py 
['fmea model', 'sys1 fail cause']
sys1 Fail mode
System2 Fail Mode
```
